src/shogun/mathematics/linalg/LinalgNamespace.h

```diff
@@ -106,6 +106,45 @@ namespace shogun
 				return sg_linalg->get_cpu_backend();
 		}
 
+		/** Infer the appropriate backend for linalg operations
+		 * from the input SGVector or SGMatrix (Container).
+		 * Raise error if the backends of the Containers conflict.
+		 *
+		 * @param a The first SGVector/SGMatrix
+		 * @param b The second SGVector/SGMatrix
+		 * @param c The third SGVector/SGMatrix
+		 * @return @see LinalgBackendBase pointer
+		 */
+		template <typename T, template <typename> class Container>
+		LinalgBackendBase* infer_backend(
```

Comment:
Can't we make this a bit nicer here, so that any number of things that have the `.on_gpu(), .on_cpu()` method implemented is checked?

@lisitsyn can you do some magic ?